crates/binstalk-downloader/src/resolver.rs

```diff
@@ -0,0 +1,75 @@
+use std::{net::SocketAddr, sync::Arc};
+
+use hyper::client::connect::dns::Name;
+use once_cell::sync::OnceCell;
+use reqwest::dns::{Addrs, Resolve};
+#[cfg(unix)]
+use trust_dns_resolver::system_conf;
+use trust_dns_resolver::{
+    config::{NameServerConfig, Protocol, ResolverConfig, ResolverOpts},
+    TokioAsyncResolver,
+};
+
+#[derive(Debug, Default, Clone)]
+pub struct DefaultResolver(Arc<OnceCell<TokioAsyncResolver>>);
+
+impl Resolve for DefaultResolver {
+    fn resolve(&self, name: Name) -> reqwest::dns::Resolving {
+        let resolver = self.clone();
+        Box::pin(async move {
+            let resolver = resolver.0.get_or_try_init(new_resolver)?;
+
+            let lookup = resolver.lookup_ip(name.as_str()).await?;
+            let addrs: Addrs = Box::new(lookup.into_iter().map(|ip| SocketAddr::new(ip, 0)));
+            Ok(addrs)
+        })
+    }
+}
+
+fn new_resolver() -> Result<TokioAsyncResolver, Box<dyn std::error::Error + Send + Sync>> {
+    #[cfg(unix)]
+    {
+        let (config, opts) = system_conf::read_system_conf()?;
+        Ok(TokioAsyncResolver::tokio(config, opts)?)
+    }
+    #[cfg(windows)]
+    {
+        let mut config = ResolverConfig::new();
+        let opts = ResolverOpts::default();
+
+        let current_interface = default_net::get_default_interface()?;
+        let adapters = ipconfig::get_adapters()?;
+        let ipaddrs = adapters
+            .iter()
+            .filter_map(|adapter| {
+                if adapter.adapter_name() == current_interface.name {
+                    Some(adapter.dns_servers())
+                } else {
+                    None
+                }
+            })
+            .flatten();
+
+        for ipaddr in ipaddrs {
+            config.add_name_server(NameServerConfig {
+                socket_addr: SocketAddr::new(ipaddr.to_owned(), 53),
+                protocol: Protocol::Tcp,
+                tls_dns_name: None,
+                trust_nx_responses: false,
+                #[cfg(feature = "rustls")]
+                tls_config: None,
+                bind_addr: None,
+            });
+            config.add_name_server(NameServerConfig {
+                socket_addr: SocketAddr::new(ipaddr.to_owned(), 53),
+                protocol: Protocol::Udp,
```

Comment:
I wonder if it is possible to also support `Protocol::{Tls, Https, Quic}`.

You can refactor this into a for loop:

```rust
for protocol in [Protocol::Quic, Protocl::Https, Protocol::Tls, Protocol::Udp, Protocol::Tcp] {
    ...
}
```

Reply:
I'm fine with doing this if you wish, but `trust-dns` doesn't do this in the default Windows resolver implementation:
https://github.com/bluejekyll/trust-dns/blob/64280f475e82d4bd940b943264eed5e85a160c65/crates/resolver/src/system_conf/windows.rs#L31-L48

Reply:
Thanks, I discovered that the unix one does the same thing

https://github.com/bluejekyll/trust-dns/blob/64280f475e82d4bd940b943264eed5e85a160c65/crates/resolver/src/system_conf/unix.rs#L66

I wonder how `trust-dns-resolver` discover and use quic, https and tls dns then...